contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,866 @@
+        require(_voteToOperator[newVoteAddress] == address(0), "DUPLICATE_VOTE_ADDRESS");
+        require(_checkVoteAddress(newVoteAddress, blsProof), "INVALID_VOTE_ADDRESS");
+
+        bytes memory oldVoteAddress = valInfo.voteAddress;
+        delete _voteToOperator[oldVoteAddress];
+        _voteToOperator[newVoteAddress] = operatorAddress;
+
+        valInfo.voteAddress = newVoteAddress;
+        valInfo.updateTime = block.timestamp;
+
+        emit VoteAddressEdited(operatorAddress, newVoteAddress);
+    }
+
+    function unjail(address operatorAddress) public onlyInitialized validatorExist(operatorAddress) {
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.jailed, "NOT_JAILED");
+
+        address pool = valInfo.poolModule;
+        require(IStakePool(pool).getSelfDelegationBNB() >= minSelfDelegationBNB, "NOT_ENOUGH_SELF_DELEGATION");
+        require(valInfo.jailUntil <= block.timestamp, "STILL_JAILED");
+
+        valInfo.jailed = false;
+        emit ValidatorUnjailed(operatorAddress);
+    }
+
+    function delegate(address operatorAddress)
+        external
+        payable
+        onlyInitialized
+        validatorExist(operatorAddress)
+        whenNotPaused
+    {
+        uint256 bnbAmount = msg.value;
+        require(bnbAmount >= minDelegationBNBChange, "INVALID_DELEGATION_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+        address pool = valInfo.poolModule;
+        require(_checkPoolImplementation(pool), "IMPLEMENTATION_NOT_MATCH");
+
+        if (valInfo.jailed || IStakePool(pool).getSelfDelegationBNB() < minSelfDelegationBNB) {
+            // only self delegation
+            require(delegator == operatorAddress, "ONLY_SELF_DELEGATION");
+        }
+
+        uint256 shares = IStakePool(pool).delegate{ value: bnbAmount }(delegator);
+        emit Delegated(operatorAddress, delegator, shares, bnbAmount);
+
+        IGovBNB(govBNB).mint(operatorAddress, delegator, bnbAmount);
+    }
+
+    function undelegate(
+        address operatorAddress,
+        uint256 shares
+    ) public onlyInitialized validatorExist(operatorAddress) whenNotPaused {
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+        require(
+            IStakePool(valInfo.poolModule).getPooledBNBByShares(shares) >= minDelegationBNBChange,
+            "INVALID_UNDELEGATION_AMOUNT"
+        );
+
+        uint256 bnbAmount = IStakePool(valInfo.poolModule).undelegate(delegator, shares);
+        emit Undelegated(operatorAddress, delegator, shares, bnbAmount);
+
+        if (
+            delegator == operatorAddress && IStakePool(valInfo.poolModule).getSelfDelegationBNB() < minSelfDelegationBNB
+        ) {
+            _validators[operatorAddress].jailed = true;
+            _removeEligibleValidator(valInfo.consensusAddress);
```

Comment:
I think it is not need to keep the `_eligibleValidators` and `_eligibleValidatorVoteAddrs`.
and the `updateEligibleValidators` should goes to `ValidatorSet` contract.